src/mcts/search.cc

```diff
@@ -458,26 +460,43 @@ EdgeAndNode Search::GetBestChildWithTemperature(Node* parent,
 
 void Search::StartThreads(size_t how_many) {
   Mutex::Lock lock(threads_mutex_);
-  while (threads_.size() < how_many) {
+  // First thread is a watchdog thread.
+  if (threads_.size() == 0) {
+    threads_.emplace_back([this]() { WatchdogThread(); });
+  }
+  // Start working threads.
+  while (threads_.size() <= how_many) {
     threads_.emplace_back([this]() {
       SearchWorker worker(this);
       worker.RunBlocking();
     });
   }
 }
 
-void Search::RunSingleThreaded() {
-  SearchWorker worker(this);
-  worker.RunBlocking();
+void Search::RunBlocking(size_t threads) {
+  StartThreads(threads);
+  Wait();
 }
 
-void Search::RunBlocking(size_t threads) {
-  if (threads == 1) {
-    RunSingleThreaded();
-  } else {
-    StartThreads(threads);
-    Wait();
+bool Search::IsSearchActive() const {
+  Mutex::Lock lock(counters_mutex_);
+  return !stop_;
+}
+
+void Search::WatchdogThread() {
+  while (IsSearchActive()) {
+    {
+      using namespace std::chrono_literals;
+      constexpr auto kMaxWaitTime = 100ms;
+      Mutex::Lock lock(counters_mutex_);
+      auto remaining_time = (limits_.time_ms - GetTimeSinceStart()) * 1ms;
```

Comment:
Should this check time_ms > 0, or do a max to ensure remaining_time is not less than 0?

Reply:
Or maybe even protect against 0 to ensure this method can't be a hard spin which dominates the lock and makes it slower for the first playout to complete in the worst case?